docs/decisions/001-strip-whitespaces.md

```diff
@@ -0,0 +1,235 @@
+  <span><span></span><span></span></span>
+
+  <!-- Don't -->
+  <span>
+    <span></span>
+    <span></span>
+  </span>
+  ```
+
+- between a tag and its text content
+
+  ```html
+  <!-- Do -->
+  <span>Hello world!</span>
+
+  <!-- Don't -->
+  <span>
+    Hello world!
+  </span>
+
+  <!-- Don't -->
+  <span> Hello world! </span>
+  ```
+
+Note: `<span />` is used in these examples, but it could be any other tag.
+
+### Concrete actions
+
+#### Use the `spaceless` tag
+
+Wrap your template within `{% spaceless %}...{% endspaceless %}`:
+
+<!-- prettier-ignore -->
+```html
+{% spaceless %}
+
+{#
+  Parameters:
+  - ...
+#}
+
+{# Pre-processing #}
+{% set ... %}
+
+{# Render #}
+<my-component>
+  ...
+</my-component>
+
+{% endspaceless %}
+```
+
+If you need to add a whitespace somewhere in your template between 2 HTML tags, you can do it with: `{% endspaceless %} {% spaceless %}`.
+
+#### Use the whitespace control modifier on your tags when needed
+
+You can trim leading and trailing whitespaces with `-` (dash) in the following ways:
+
+- `{{- my_var }}`: remove left whitespace
+- `{{ my_var -}}`: remove right whitespace
+- `{{- my_var -}}`: remove both whitespaces
+
+##### Simple scenario: the content is surrounded by HTML tags
+
+It makes sense to use it when you print content between 2 tags:
+
+```html
+<!-- Do -->
+</span>
+  {{- my_var -}}
+<span>
+
+<!-- Don't -->
+</span>
+  {{ my_var }}
+<span>
+```
+
+or inside a tag:
+
+```html
+<!-- Do -->
+<span>
+  {{- my_var -}}
+</span>
+
+<!-- Don't -->
+<span>
+  {{ my_var }}
+</span>
+```
+
+But it doesn’t make sense in attributes,
+
+```html
+<!-- Do -->
+<span {{ my_var }} />
+
+<!-- Don't -->
+<span {{- my_var -}} />
+```
+
+or when there’s no space around:
+
+```html
+<!-- Do -->
+</span>{{ my_var }}<span>
+
+<!-- Don't -->
+</span>{{- my_var -}}<span>
+```
+
+```html
+<!-- Do -->
+<span>{{ my_var }}</span>
+
+<!-- Don't -->
+<span>{{- my_var -}}</span>
+```
+
+Note: `{{` can be replaced by any content-printing structure, e.g. `{% include '...' %}`, `{% embed '...' %}` or `{% block '...' %}`.
+
+##### Complex scenario: the content is surrounded by Twig tags
+
+Sometimes, you will find that control structures (`if`, `for`) and other Twig tags (like `set` or comments) may alter the output by introducing undesired whitespaces.
+
+As a rule of thumb, it's good to start by removing all the whitespaces for these tags using the dash syntax, generate the snapshot of the finished component and then start removing the dash where they're unnecessary.
+
+Here's the plan to follow:
+
+1. the developper creates the component X with dashes on all the control structures and other tags that might generate whitespaces
+2. the developper takes the Jest snapshots and open a PR
+3. the reviewer validates the component and the snapshots, and then merges the PR
+4. the reviewer creates a Jira ticket: "Improve quality of component X". The goal of the ticket is to remove as many dashes as possible.
```

Comment:
If i get it correctly the idea would be to apply blindly dashes to remove any potential whitespace, merge if the component looks god and after that creating a follow up ticket to verify if there's any chance to remove dashes here and there.
I haven't fully partecipated to the elaboration of this helpful document, but this flow looks inefficient to me, why don't we consider the usage of whitespace removal as any other "syntax" related thing? Ideally it should come as a "prettier" option, although it is something that might needs a case by case evaluation.

Reply:
the issue is that there is no simple rule to determine if we should put a dash.
Sometime you have to add it to a `if`, sometimes you don't

Reply:
We can also write the whole code on a single line: this way, we are sure the code we produce is ok. React does it for us already. If we wanted to be 100% accurate on the ECL website, we should tell our beautifier to not add any newline and to not indent the code displayed in the showcase... but then, it wouldn't be really readable. 

Prettier describes it well I think: https://prettier.io/blog/2018/11/07/1.15.0.html#whitespace-sensitive-formatting 

The code produced by the twig template must work as expected. It should not introduce breaking whitespaces, as it was the case for the link component. The only way we have to ensure this is to remove all the whitespaces between an within HTML tags.

I hope this helps clarifying the context :)

This question of whitespaces is not only a matter of "syntax", it's not as choosing tabs vs spaces. 

`spaceless` is pretty good at removing whitespaces between HTML tags. The main issue we face is when we mix HTML tags and text. That's when things go bad...